### PR TITLE
use more of QColor functions / methods

### PR DESCRIPTION
**Pull request by pijyoi**
This PR makes more use of QColor functions / methods
- static QColor.fromXXX()
- QColor.getRgbF()
- QColor.name()

Several helper functions in pg.functions are effectively trivial reimplementations of the above functions / methods
- class Color
- colorTuple
- colorStr
- glColor

The following line was left unconverted because it seems that the format string does not match the data types. (%f to integer, %d to float [0, 1])
https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/exporters/SVGExporter.py#L122

### CODE REVIEW

File: pyqtgraph/colormap.py [hunk 1]
```diff
@@ -606,9 +604,9 @@ def map(self, data, mode=BYTE):
         # Convert to QColor if requested
         if mode == self.QCOLOR:
             if np.isscalar(data):
-                return QtGui.QColor(*interp)
+                return QtGui.QColor.fromRgbF(*interp)
             else:
-                return [QtGui.QColor(*x.tolist()) for x in interp]
+                return [QtGui.QColor.fromRgbF(*x.tolist()) for x in interp]
         else:
```
Comment (NilsNemitz) [inline pyqtgraph/colormap.py:610]: Thank you, that is very much what I had in mind!

I am glad that you implemented this; I am sorry that I didn't find any time to add this after you pointed out the factory functions.

File: pyqtgraph/colormap.py [hunk 2]
```diff
@@ -762,8 +760,10 @@ def getStops(self, mode=BYTE):
                 color = color.astype(float) / 255.
             elif mode == self.QCOLOR:
                 if color.dtype.kind == 'f':
-                    color = (color*255).astype(np.ubyte)
-                color = [QtGui.QColor(*x.tolist()) for x in color]
+                    factory = QtGui.QColor.fromRgbF
+                else:
+                    factory = QtGui.QColor.fromRgb
+                color = [factory(*x.tolist()) for x in color]
```
Comment (NilsNemitz) [inline pyqtgraph/colormap.py:766]: neat :)

File: pyqtgraph/graphicsItems/PlotItem/PlotItem.py [hunk 3]
```diff
@@ -763,9 +763,8 @@ def writeSvgCurves(self, fileName=None):
 
             for item in self.curves:
                 if isinstance(item, PlotCurveItem):
-                    color = fn.colorStr(item.pen.color())
-                    opacity = item.pen.color().alpha() / 255.
-                    color = color[:6]
+                    color = item.pen.color()
+                    hrrggbb, opacity = color.name(), color.alphaF()
```
Comment (NilsNemitz) [inline pyqtgraph/graphicsItems/PlotItem/PlotItem.py:767]: Out of curiosity, what does the `h` in `hrrggbb` stand for?

Comment (pijyoi) [reply pyqtgraph/graphicsItems/PlotItem/PlotItem.py:767]: hex or hash. The name() method returns a str of the form '#rrggbb'. It's to remind that a hex or hash has already been prepended.

File: pyqtgraph/opengl/items/GLMeshItem.py [hunk 4]
```diff
@@ -181,7 +181,7 @@ def paint(self):
                     if self.colors is None:
                         color = self.opts['color']
                         if isinstance(color, QtGui.QColor):
-                            glColor4f(*fn.glColor(color))
+                            glColor4f(*color.getRgbF())
```
Comment (NilsNemitz) [inline pyqtgraph/opengl/items/GLMeshItem.py:184]: As the code checker points out, it seems you have successfully eliminated the dependency on pg.functions !

Comment (pijyoi) [reply pyqtgraph/opengl/items/GLMeshItem.py:184]: There are inconsistencies with the opengl items' APIs. Some take in float colors, some take in an argument for mkColor.

**Comment by j9ac9k** [comment]
@pijyoi thank you for this!  Been thinking there was likely room to implement the QColor methods a bit more directly.

The one gotcha, which it looks like you've taken care to avoid, is that `QColor::setNamedColor` wants `#AARRGGBB` (if you're going to specify alpha), whereas pyqtgraph expects `#RRGGBBAA`.

As you pointed out, we have some inconsistencies across the board for the opengl item APIs; not sure if trying to modify those would be part of this PR or not (I'm not opposed to it).

**Comment by j9ac9k** [comment]
Also I'm totally good with putting deprecation warnings (or should we be putting future warnings, I'm always uncertain) on some of the stuff that's rendered more or less useful (like `Color` class, `colorTuple`, `colorStr` and `glColor`).

**Comment by NilsNemitz** [comment]
@pijyoi , thank you for this clean-up! I also see this as a great improvement, since it removes a lot of redundancy and alternative definitions in favor of simply using the Qt conventions.

Fortunately it turns out that the data formats are mostly compatible; Much more so than I would have expected. That makes the now redundant function definitions so simple that there is not much cost* to keeping them. They actually serve as excellent demonstrations of direct replacement code - for those who look at the source.
(* for "cost" I would consider potential for user and developer confusion most important; this seems near zero here.)

My suggestion would be to take a quick look over the documentation for the colors:
The first step to getting the legacy functions out of user code would be to no longer advertise them, and to make sure that alternatives are easy to find... This might include some hint towards the (somewhat hidden) Qt factory functions.

A simple approach might be to move the now-redundant set of functions to a "legacy functions" section of the page for now and basically copy-and-paste in an example of the equivalent Qt dialect as a suggested replacement?

**Comment by pijyoi** [comment]
>As you pointed out, we have some inconsistencies across the board for the opengl item APIs; not sure if trying to modify those would be part of this PR or not (I'm not opposed to it).

From what I can see, the various items can be grouped thus:

passed to mkColor
- GLBoxItem
- GLGridItem
- GLTextItem (although only documented to take an array of R,G,B,A)

2D-ndarray of float colors if different color for each element
4-tuple of float or QColor if single color for every element
- GLMeshItem
- GLScatterPlotItem
- GLSurfacePlotItem
- GLLinePlotItem (in addition takes a str passed to mkColor for scalar (undocumented))

GLGradientLegendItem
- fontColor : 4-tuple of float only

The only glaring inconsistencies are:
- GLGradientLegendItem. This looks like a bug. It should be changed to take ```mkColor()```
   * Unless you consider GLGradientLegendItem to be "correct" and GLBoxItem, GLGridItem, GLTextItem to be "wrong"
- GLLinePlotItem taking a ```str``` overload. Probably should just leave it as is.

**Comment by NilsNemitz** [comment]
As stated, I have some tolerance for just keeping the old definitions around without aggressive deprecation.

But I do believe we should deprecate the `#RRGGBBAA` format as soon as possible. There are many other options to specify alpha, and I really hope that this is not so commonly used.

Once we have officially removed this usage, we can re-add the Qt compatible `#AARRGGBB`. 
I am sure there are good reasons for one or the other, but after this PR is merged, PyQtGraph will do color "the Qt way" basically everywhere. And that places us -at least from my point of view- very firmly on the `#AARRGGBB` side of the fence.

It would also remove at least one case distinctions from the painfully long list in mkColor. 🤯

**Comment by j9ac9k** [comment]
> Once we have officially removed this usage, we can re-add the Qt compatible `#AARRGGBB`.
> I am sure there are good reasons for one or the other, but after this PR is merged, PyQtGraph will do color "the Qt way" basically everywhere. And that places us -at least from my point of view- very firmly on the `#AARRGGBB` side of the fence.

This change is a tough sell IMO.  There is no good way of rolling out this change without silently breaking code for existing users.  I suppose we could have that conversation if we threw in a long enough of a FutureWarning when passing hex values of RGBA, but I think at this point we should likely keep the hex ordering consistent with the library.

File: pyqtgraph/opengl/items/GLGradientLegendItem.py [hunk 5]
```diff
@@ -18,7 +18,8 @@ def __init__(self, **kwds):
             gradient: a pg.ColorMap used to color the colorbar
             labels: a dict of "text":value to display next to the colorbar.
                 The value corresponds to a position in the gradient from 0 to 1.
-            fontColor: sets the color of the texts
+            fontColor: sets the color of the texts.
+                Arguments are the same as those accepted by functions.mkColor().
```
Comment (j9ac9k) [inline pyqtgraph/opengl/items/GLGradientLegendItem.py:22]: To do a link within this docs, you can change to

```python
:func:`~pyqtgraph.mkColor`
```

**Comment by pijyoi** [comment]
I agree with @NilsNemitz that the existing trivial functions / classes can be left behind without immediate deprecation.

I am not so sure about the #AARRGGBB vs #RRGGBBAA. #AARRGGBB feels very much like we are getting exposed to the internal Qt representation. i.e. alpha in MSB and blue in LSB of a 32-bit word.
```python
In [6]: hex(QtGui.QColor(0x33, 0x66, 0x99).rgba())
Out[6]: '0xff336699'
```

**Comment by NilsNemitz** [comment]
@j9ac9k , I *was* advocating a deprecation period. :)

Maybe a compromise would be to have the documentation and docstrings state that 4-byte hex-strings are not ideal for unambiguously giving RGB + alpha, since the byte ordering is less well-defined as for RGB. That might make it easier to switch in the future, and doesn't really hurt anything in the present.

**Comment by j9ac9k** [comment]
> @j9ac9k , I _was_ advocating a deprecation period. :)
> 
> Maybe a compromise would be to have the documentation and docstrings state that 4-byte hex-strings are not ideal for unambiguously giving RGB + alpha, since the byte ordering is less well-defined as for RGB. That might make it easier to switch in the future, and doesn't really hurt anything in the present.

I'd be good with that; At the very least we should explicitly state that we do not follow the Qt API.

@pijyoi The benefit of adopting `#AARRGGBB` would be not for the sake of the hex conversion, but the code in `mkColor` can be simplified in a way such that...

```python
if isinstance(arg, str):
    return QColor.setNamedColor(arg)
```

That would give us the benefit of not just being able to offload the hex logic (which isn't particularly complex); but also lets us access the [SVG color keyword names](https://www.w3.org/TR/SVG11/types.html#ColorKeywords); which I think would be useful.

Granted, we can easily accommodate both (get the color by SVG name, and handle our own hex string ordering).

File: pyqtgraph/opengl/GLViewWidget.py [hunk 6]
```diff
@@ -138,9 +138,9 @@ def initializeGL(self):
     def setBackgroundColor(self, *args, **kwds):
         """
         Set the background color of the widget. Accepts the same arguments as
-        pg.mkColor() and pg.glColor().
+        pg.mkColor().
```
Comment (j9ac9k) [inline pyqtgraph/opengl/GLViewWidget.py:141]: I know this wasn't placed there before, but we can link the function here too:

```
:func:`~pyqtgraph.mkColor`
```

**Comment by pijyoi** [comment]
>That would give us the benefit of not just being able to offload the hex logic (which isn't particularly complex); but also lets us access the SVG color keyword names; which I think would be useful.

I added it in as an experimental feature. Perhaps to become officially documented when the deprecated non-#-prefixed hex strings gets removed in 0.13?

```python
In [4]: pg.mkColor('tomato').getRgb()
Out[4]: (255, 99, 71, 255)
```

**Comment by j9ac9k** [comment]
> I added it in as an experimental feature. Perhaps to become officially documented when the deprecated non-#-prefixed hex strings gets removed in 0.13?

That's a great idea.

**Comment by j9ac9k** [comment]
@pijyoi this LGTM, I see no problem merging this right now.  Are you planning on adding anything else to this PR?

**Comment by pijyoi** [comment]
> Are you planning on adding anything else to this PR?

No

**Comment by j9ac9k** [comment]
Ha, oddly enough while working on that IEEE paper, I think I came across that SVG background bug, not sure why I didn't write up an issue on that given I opened numerous other issues that I encountered while generating plots for that... oh well.

This LGTM, thanks so much @pijyoi the code is so much simpler in so many places!

**Pull request merged**